impl/src/main/java/org/jboss/weld/event/ObserverNotifier.java

```diff
@@ -352,11 +358,35 @@ public RuntimeException apply(Type eventType) {
             }), executor);
         }
 
-        // TODO if timeout set then wrap the completableFuture
-
+        // If NotificationOptionKeys.TIMEOUT is set, we will trigger the counter and use CompletableFuture.anyOf()
+        if (timeout != null) {
+            completableFuture = CompletableFuture.anyOf(completableFuture, startTimer(timeout)).thenApply((ignoredObject) -> event);
+        }
         return new AsyncEventDeliveryStage<>(completableFuture, executor);
     }
 
+    /**
+     * Verifies that, if timeout options was set, the executor is available and input value for timeout can be interpreted as Long.
+     * Returns the timeout value if all is alright, null if this option was not requested.
+     */
+    private Long verifyNotificationTimeoutOption(NotificationOptions options) {
```

Comment:
If you pass the timeout option value you can save one `options.get()` invocation. I know it's a supermicro-optimization but I would like to keep the init/validation methods as similar as possible.